server/drafty/drafty.go

```diff
@@ -231,8 +233,102 @@ func toTree(drafty *document) (*node, error) {
 	return &node{children: children}, nil
 }
 
+// given a grapheme iterator, start and end pos, returns a new grapheme iterator
+// containing a slice of graphemes(start->end) from input interator
+func sliceGraphemeClusters(g *uniseg.Graphemes, start, end int) *uniseg.Graphemes {
+	if g == nil {
+		return uniseg.NewGraphemes("")
+	}
+
+	g.Reset()
+
+	i := 0
+	j := -1
+
+	output := ""
+
+	for g.Next() {
+		if j > 0 {
+			if j < end {
+				output = output + g.Str()
+				j++
+			} else {
+				// end pos found, stop collecting string
+				break
+			}
+		} else if i < start {
+			i++
+		} else if i == start {
+			// starting pos found, start collecting string
+			output = output + g.Str()
+			j = i + 1
+		}
+
+	}
+
+	g.Reset()
+
+	return uniseg.NewGraphemes(output)
+}
+
+// given a grapheme iterator, returns the original string from which it was created from
+func graphemeToString(g *uniseg.Graphemes) string {
+	if g == nil {
```

Comment:
How can it be `nil`?

Reply:
Yes, you are right, current code scenarios it is never assigned to nil, but since it is a pointer and somehow got assigned to nil, g.Next() would panic.

What I can do is return error if g is nil